modules/dns/dns_api.rb

```diff
@@ -27,17 +39,41 @@ class Api < ::Sinatra::Base
       end
     end
 
-    delete "/:value" do
-      type = params[:value] =~ /\.(in-addr|ip6)\.arpa$/ ? "PTR" : "A"
+    delete '/:value/?:type?' do
+      validate_dns_name(params[:value])
+
+      if params[:type]
+        type = params[:type]
+      else
+        type = params[:value] =~ /\.(in-addr|ip6)\.arpa$/ ? "PTR" : "A"
+      end
 
       begin
-        server.remove_a_record(params[:value]) if type == 'A'
-        server.remove_ptr_record(params[:value]) if type == 'PTR'
+        case type
+        when 'A'
+          server.remove_a_record(params[:value])
+        when 'AAAA'
+          server.remove_aaaa_record(params[:value])
+        when 'PTR'
+          validate_reverse_dns_name(params[:value])
+          server.remove_ptr_record(params[:value])
+        else
+          log_halt(400, "unrecognized 'type' parameter: #{type}")
+        end
       rescue Proxy::Dns::NotFound => e
         log_halt 404, e
       rescue => e
         log_halt 400, e
       end
     end
+
+    def validate_dns_name name
+      log_halt(400, "Invalid DNS name #{name}") unless name =~ /^([a-zA-Z0-9]([-a-zA-Z0-9]+)?\.?)+$/
```

Comment:
pls. don't perform validation and handling of failed validation in the same method.